File: ruby/test/basecamp/chain_hooks_test.rb
# frozen_string_literal: true

require "test_helper"

class ChainHooksTest < Minitest::Test
  def test_calls_start_hooks_in_order
    calls = []
    hook_a = RecordingHooks.new("A", calls)
    hook_b = RecordingHooks.new("B", calls)
    chain = Basecamp::ChainHooks.new(hook_a, hook_b)

    info = Basecamp::RequestInfo.new(method: "GET", url: "/test")
    chain.on_request_start(info)

    assert_equal [ [ "A", :on_request_start ], [ "B", :on_request_start ] ], calls
  end

  def test_calls_end_hooks_in_reverse_order
    calls = []
    hook_a = RecordingHooks.new("A", calls)
    hook_b = RecordingHooks.new("B", calls)
    chain = Basecamp::ChainHooks.new(hook_a, hook_b)

    info = Basecamp::RequestInfo.new(method: "GET", url: "/test")
    result = Basecamp::RequestResult.new(status_code: 200)
    chain.on_request_end(info, result)

    assert_equal [ [ "B", :on_request_end ], [ "A", :on_request_end ] ], calls
  end

  def test_calls_operation_start_in_order
    calls = []
    hook_a = RecordingHooks.new("A", calls)
    hook_b = RecordingHooks.new("B", calls)
    chain = Basecamp::ChainHooks.new(hook_a, hook_b)

    info = Basecamp::OperationInfo.new(service: "projects", operation: "list")
    chain.on_operation_start(info)

    assert_equal [ [ "A", :on_operation_start ], [ "B", :on_operation_start ] ], calls
  end

  def test_calls_operation_end_in_reverse_order
    calls = []
    hook_a = RecordingHooks.new("A", calls)
    hook_b = RecordingHooks.new("B", calls)
    chain = Basecamp::ChainHooks.new(hook_a, hook_b)

    info = Basecamp::OperationInfo.new(service: "projects", operation: "list")
    result = Basecamp::OperationResult.new(duration_ms: 100)
    chain.on_operation_end(info, result)

    assert_equal [ [ "B", :on_operation_end ], [ "A", :on_operation_end ] ], calls
  end

  def test_calls_retry_hooks_in_order
    calls = []
    hook_a = RecordingHooks.new("A", calls)
    hook_b = RecordingHooks.new("B", calls)
    chain = Basecamp::ChainHooks.new(hook_a, hook_b)

    info = Basecamp::RequestInfo.new(method: "GET", url: "/test")
    chain.on_retry(info, 2, StandardError.new, 1.0)

    assert_equal [ [ "A", :on_retry ], [ "B", :on_retry ] ], calls
  end

  def test_calls_paginate_hooks_in_order
    calls = []
    hook_a = RecordingHooks.new("A", calls)
    hook_b = RecordingHooks.new("B", calls)
    chain = Basecamp::ChainHooks.new(hook_a, hook_b)

    chain.on_paginate("/url", 1)

    assert_equal [ [ "A", :on_paginate ], [ "B", :on_paginate ] ], calls
  end

  def test_swallows_exceptions_from_hooks
    bad_hook = ExplodingHooks.new
    calls = []
    good_hook = RecordingHooks.new("good", calls)
    chain = Basecamp::ChainHooks.new(bad_hook, good_hook)

    info = Basecamp::RequestInfo.new(method: "GET", url: "/test")
    chain.on_request_start(info)

    # The good hook was still called despite the bad hook raising
    assert_equal [ [ "good", :on_request_start ] ], calls
  end

  def test_swallows_exceptions_on_end_hooks
    calls = []
    good_hook = RecordingHooks.new("good", calls)
    bad_hook = ExplodingHooks.new
    chain = Basecamp::ChainHooks.new(good_hook, bad_hook)

    info = Basecamp::RequestInfo.new(method: "GET", url: "/test")
    result = Basecamp::RequestResult.new(status_code: 200)
    # on_request_end calls in reverse: bad_hook first (raises), then good_hook
    chain.on_request_end(info, result)

    assert_equal [ [ "good", :on_request_end ] ], calls
  end

  private

  class RecordingHooks
    include Basecamp::Hooks

    def initialize(name, calls)
      @name = name
      @calls = calls
    end

    def on_operation_start(_info)
      @calls << [ @name, :on_operation_start ]
    end

    def on_operation_end(_info, _result)
      @calls << [ @name, :on_operation_end ]
    end

    def on_request_start(_info)
      @calls << [ @name, :on_request_start ]
    end

    def on_request_end(_info, _result)
      @calls << [ @name, :on_request_end ]
    end

    def on_retry(_info, _attempt, _error, _delay)
      @calls << [ @name, :on_retry ]
    end

    def on_paginate(_url, _page)
      @calls << [ @name, :on_paginate ]
    end
  end

  class ExplodingHooks
    include Basecamp::Hooks

    def on_request_start(_info)
      raise "boom"
    end

    def on_request_end(_info, _result)
      raise "boom"
    end

    def on_operation_start(_info)
      raise "boom"
    end

    def on_operation_end(_info, _result)
      raise "boom"
    end
  end
end

File: kotlin/sdk/src/commonTest/kotlin/com/basecamp/sdk/ChainHooksTest.kt
package com.basecamp.sdk

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.time.Duration.Companion.milliseconds

class ChainHooksTest {

    private val info = OperationInfo(
        service = "Projects",
        operation = "ListProjects",
        resourceType = "project",
        isMutation = false,
    )

    private val requestInfo = RequestInfo(method = "GET", url = "https://example.com", attempt = 1)

    @Test
    fun startEventsFireInForwardOrder() {
        val order = mutableListOf<String>()

        val hook1 = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                order.add("hook1")
            }
        }
        val hook2 = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                order.add("hook2")
            }
        }
        val hook3 = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                order.add("hook3")
            }
        }

        val chain = chainHooks(hook1, hook2, hook3)
        chain.onOperationStart(info)

        assertEquals(listOf("hook1", "hook2", "hook3"), order)
    }

    @Test
    fun endEventsFireInReverseOrder() {
        val order = mutableListOf<String>()
        val result = OperationResult(duration = 100.milliseconds)

        val hook1 = object : BasecampHooks {
            override fun onOperationEnd(info: OperationInfo, result: OperationResult) {
                order.add("hook1")
            }
        }
        val hook2 = object : BasecampHooks {
            override fun onOperationEnd(info: OperationInfo, result: OperationResult) {
                order.add("hook2")
            }
        }
        val hook3 = object : BasecampHooks {
            override fun onOperationEnd(info: OperationInfo, result: OperationResult) {
                order.add("hook3")
            }
        }

        val chain = chainHooks(hook1, hook2, hook3)
        chain.onOperationEnd(info, result)

        assertEquals(listOf("hook3", "hook2", "hook1"), order)
    }

    @Test
    fun requestStartForwardRequestEndReverse() {
        val startOrder = mutableListOf<String>()
        val endOrder = mutableListOf<String>()
        val requestResult = RequestResult(statusCode = 200, duration = 50.milliseconds)

        val hook1 = object : BasecampHooks {
            override fun onRequestStart(info: RequestInfo) {
                startOrder.add("hook1")
            }
            override fun onRequestEnd(info: RequestInfo, result: RequestResult) {
                endOrder.add("hook1")
            }
        }
        val hook2 = object : BasecampHooks {
            override fun onRequestStart(info: RequestInfo) {
                startOrder.add("hook2")
            }
            override fun onRequestEnd(info: RequestInfo, result: RequestResult) {
                endOrder.add("hook2")
            }
        }

        val chain = chainHooks(hook1, hook2)
        chain.onRequestStart(requestInfo)
        chain.onRequestEnd(requestInfo, requestResult)

        assertEquals(listOf("hook1", "hook2"), startOrder)
        assertEquals(listOf("hook2", "hook1"), endOrder)
    }

    @Test
    fun emptyChainReturnsNoopHooks() {
        val chain = chainHooks()
        assertEquals(NoopHooks, chain)
        // Should not throw
        chain.onOperationStart(info)
        chain.onOperationEnd(info, OperationResult(duration = 0.milliseconds))
    }

    @Test
    fun singleHookReturnedDirectly() {
        val hook = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {}
        }
        val chain = chainHooks(hook)
        assertEquals(hook, chain, "Single hook should be returned directly, not wrapped")
    }

    @Test
    fun noopHooksAreFiltered() {
        val order = mutableListOf<String>()
        val hook1 = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                order.add("hook1")
            }
        }

        val chain = chainHooks(NoopHooks, hook1, NoopHooks)
        // With only one active hook, chainHooks returns it directly
        assertEquals(hook1, chain)
    }

    @Test
    fun retryHookFiresInForwardOrder() {
        val order = mutableListOf<String>()
        val error = RuntimeException("test")

        val hook1 = object : BasecampHooks {
            override fun onRetry(info: RequestInfo, attempt: Int, error: Throwable, delayMs: Long) {
                order.add("hook1")
            }
        }
        val hook2 = object : BasecampHooks {
            override fun onRetry(info: RequestInfo, attempt: Int, error: Throwable, delayMs: Long) {
                order.add("hook2")
            }
        }

        val chain = chainHooks(hook1, hook2)
        chain.onRetry(requestInfo, 2, error, 1000L)

        assertEquals(listOf("hook1", "hook2"), order)
    }

    @Test
    fun hookExceptionsDoNotPropagate() {
        val order = mutableListOf<String>()

        val throwingHook = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                order.add("throwing")
                throw RuntimeException("hook failure")
            }
        }
        val goodHook = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                order.add("good")
            }
        }

        val chain = chainHooks(throwingHook, goodHook)
        // Should not throw even though first hook throws
        chain.onOperationStart(info)

        assertEquals(listOf("throwing", "good"), order)
    }

    @Test
    fun hookReceivesCorrectOperationInfo() {
        var capturedInfo: OperationInfo? = null

        val hook = object : BasecampHooks {
            override fun onOperationStart(info: OperationInfo) {
                capturedInfo = info
            }
        }

        val testInfo = OperationInfo(
            service = "Todos",
            operation = "CompleteTodo",
            resourceType = "todo",
            isMutation = true,
            projectId = 42,
            resourceId = 100,
        )

        val chain = chainHooks(hook)
        chain.onOperationStart(testInfo)

        val captured = capturedInfo!!
        assertEquals("Todos", captured.service)
        assertEquals("CompleteTodo", captured.operation)
        assertEquals(true, captured.isMutation)
        assertEquals(42L, captured.projectId)
        assertEquals(100L, captured.resourceId)
    }
}
